### PR TITLE
#17768: Float32 support for Inference mode in Batch Norm

### PR DESCRIPTION
**Pull request by VirdhatchaniKN**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/17768

### Problem description
To Provide Fp32 support for Inference mode of BN

### What's changed
Support provided for fp32 data type for inference mode of BN

### Checklist
- [x] [All post-commit tests](https://github.com/tenstorrent/tt-metal/actions/runs/13217558701)
- [x] [Blackhole post-commit tests](https://github.com/tenstorrent/tt-metal/actions/runs/13157671059)
- [x] [(Single-card) Tests for new models](https://github.com/tenstorrent/tt-metal/actions/runs/13217560775) - Passed as in main
- [x] [(Single-card) Demo tests](https://github.com/tenstorrent/tt-metal/actions/runs/13217560090) - Passed as in main
- [x] [(Single-card) Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/runs/13217559606)
- [x] [(Single-card) Model perf tests](https://github.com/tenstorrent/tt-metal/actions/runs/13217559245) - Passed as in main

### CODE REVIEW

**Review by umadevimcw — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/compute/batch_norm_sfpu_kernel.cpp [hunk 1]
```diff
@@ -0,0 +1,244 @@
+// SPDX-FileCopyrightText: © 2025 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "compute_kernel_api/eltwise_binary_sfpu.h"
+#include "cpp/ttnn/deprecated/tt_dnn/kernels/compute/moreh_common.hpp"
+#include "compute_kernel_api/eltwise_unary/sfpu_split_includes.h"
+#include "compute_kernel_api/eltwise_unary/eltwise_unary.h"
+#include "dprint.h"
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/compute/batch_norm_sfpu_kernel.cpp:9]: comment

**Review by KalaivaniMCW — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/eltwise/backward/utility_funcs.py [hunk 2]
```diff
@@ -18,20 +18,26 @@ def data_gen_with_range_batch_norm(
...
     assert high > low, "Incorrect range provided"
     torch.manual_seed(213919)
     channels = input_shapes[1]
     size = input_shapes if is_input else channels
-    pt_tensor = torch.rand(size, requires_grad=required_grad).bfloat16() * (high - low) + low
+    torch_dtype = torch.bfloat16
+    ttnn_dtype = ttnn.bfloat16
+    if testing_dtype == "float32":
+        torch_dtype = torch.float32
+        ttnn_dtype = ttnn.float32
```
Comment (patrickroberts) [inline tests/ttnn/unit_tests/operations/eltwise/backward/utility_funcs.py:31]:
```suggestion
    torch_dtype = getattr(torch, testing_dtype)
    ttnn_dtype = getattr(ttnn, testing_dtype)
```

File: tests/ttnn/unit_tests/operations/test_batch_norm.py [hunk 3]
```diff
@@ -10,6 +10,127 @@
...
+    bias_tt = ttnn.from_torch(bias_torch, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device) if bias else None
+
+    result_tt = ttnn.batch_norm(
+        input_tensor_tt, running_mean=batch_mean_tt, running_var=batch_var_tt, eps=eps, weight=weight_tt, bias=bias_tt
+    )
+    tt_out = ttnn.to_torch(result_tt)
+
+    status_1 = torch.allclose(result_torch, tt_out, atol=1e-10, rtol=1e-5)
+    status_2 = compare_results_batch_norm([result_torch], [tt_out])
+    assert status_2 & status_1
```
Comment (patrickroberts) [inline tests/ttnn/unit_tests/operations/test_batch_norm.py:57]:
```suggestion
    assert status_2 and status_1
```

File: tests/ttnn/unit_tests/operations/test_batch_norm.py [hunk 4]
```diff
@@ -10,6 +10,127 @@
...
+    torch_result = torch.nn.functional.batch_norm(
+        input=in_data,
+        running_mean=mean_data,
+        running_var=var_data,
+        weight=weight_data,
+        bias=bias_data,
+        training=training,
+        eps=eps,
+    )
+    comp_pass = compare_results_batch_norm([tt_output], [torch_result]) & torch.allclose(
```
Comment (patrickroberts) [inline tests/ttnn/unit_tests/operations/test_batch_norm.py:130]:
```suggestion
    comp_pass = compare_results_batch_norm([tt_output], [torch_result]) and torch.allclose(
```

File: tests/ttnn/unit_tests/operations/test_batch_norm.py [hunk 5]
```diff
@@ -10,6 +10,127 @@
...
+        (False, False, False, "float32"),  # xfail case
+        (False, True, False, "float32"),  # xfail case
+        (False, False, True, "float32"),  # xfail case
+        (False, True, True, "float32"),
+    ],
+)
+@pytest.mark.parametrize("weight", [True, False])
+@pytest.mark.parametrize("bias", [True, False])
+@pytest.mark.parametrize("eps", [1.0, 0.0, 2.34, 1e-05])
+def test_batch_norm_fp32(input_shapes, training, check_mean, check_var, weight, bias, eps, testing_dtype, device):
```
Comment (patrickroberts) [inline tests/ttnn/unit_tests/operations/test_batch_norm.py:83]: Can we just remove this `training` parameter from the PR if we're not going to test it? then just set hardcoded `training=False` in the `ttnn.batch_norm` and `torch.nn.functional.batch_norm` calls below.

Comment (patrickroberts) [reply tests/ttnn/unit_tests/operations/test_batch_norm.py:83]: Also remove `testing_dtype` and just hardcode that to `"float32"` in the calls to `data_gen_with_range_batch_norm` below. The test is called `test_batch_norm_fp32` after all.

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/dataflow/reader_batch_norm.cpp [hunk 6]
```diff
@@ -38,7 +38,13 @@ void kernel_main() {
     constexpr auto cb_id_eps = tt::CBIndex::c_4;
 
     cb_reserve_back(cb_id_eps, onetile);
-    fill_with_val_bfloat16(cb_id_eps, eps);
+#ifdef FILL_WITH_VALUE_FLOAT
+    float* float_ptr = reinterpret_cast<float*>(const_cast<long unsigned int*>(&eps));
+    FILL_WITH_VALUE_FLOAT(cb_id_eps, *float_ptr);
```
Comment (patrickroberts) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/dataflow/reader_batch_norm.cpp:43]: Please *never* use `const_cast`. This should just be
```suggestion
    FILL_WITH_VALUE_FLOAT(cb_id_eps, *reinterpret_cast<const float*>(&eps));
```

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/batch_norm_program_factory.cpp [hunk 7]
```diff
@@ -218,38 +221,80 @@ BatchNormOperation::BatchNormFactory::cached_program_t BatchNormOperation::Batch
...
+    }
+
     std::vector<uint32_t> compute_kernel_args = {
         static_cast<uint32_t>(weight_has_value), static_cast<uint32_t>(bias_has_value)};
     auto compute_kernel_id = tt_metal::CreateKernel(
         program,
-        "ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/compute/batch_norm_kernel.cpp",
+        fp32_dest_acc_en
+            ? "ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/compute/batch_norm_sfpu_kernel.cpp"
+            : "ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/compute/batch_norm_kernel.cpp",
```
Comment (patrickroberts) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/device/batch_norm_program_factory.cpp:292]: Maybe consider this?
```suggestion
        fmt::format("ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/compute/batch_norm_{}.cpp", fp32_dest_acc_en ? "sfpu_kernel" : "kernel"),
```

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/batch_norm_program_factory.cpp [hunk 8]
```diff
@@ -218,38 +221,80 @@ BatchNormOperation::BatchNormFactory::cached_program_t BatchNormOperation::Batch
...
+    if (fp32_dest_acc_en) {
+        unpack_to_dest_mode[src_input_cb_index] = UnpackToDestMode::UnpackToDestFp32;
+        unpack_to_dest_mode[src_batch_mean_cb_index] = UnpackToDestMode::UnpackToDestFp32;
+        unpack_to_dest_mode[src_batch_var_cb_index] = UnpackToDestMode::UnpackToDestFp32;
+        unpack_to_dest_mode[src_temp_num_cb_index] = UnpackToDestMode::UnpackToDestFp32;
+        unpack_to_dest_mode[src_temp_den_cb_index] = UnpackToDestMode::UnpackToDestFp32;
+        unpack_to_dest_mode[src_eps_cb_index] = UnpackToDestMode::UnpackToDestFp32;
+        unpack_to_dest_mode[src_weight_cb_index] = UnpackToDestMode::UnpackToDestFp32;
+        unpack_to_dest_mode[src_temp_1_cb_index] = UnpackToDestMode::UnpackToDestFp32;
+        unpack_to_dest_mode[src_bias_cb_index] = UnpackToDestMode::UnpackToDestFp32;
```
Comment (patrickroberts) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/device/batch_norm_program_factory.cpp:283]:
```suggestion
        for (const auto cb_index : {src_input_cb_index, src_batch_mean_cb_index, src_batch_var_cb_index, src_temp_num_cb_index, src_temp_den_cb_index, src_eps_cb_index, src_weight_cb_index, src_temp_1_cb_index, src_bias_cb_index}) {
          unpack_to_dest_mode[cb_index] = UnpackToDestMode::UnpackToDestFp32;
        }
```

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/batch_norm_device_operation.cpp [hunk 9]
```diff
@@ -8,21 +8,36 @@
...
+    TT_FATAL(
+        tensor.get_dtype() == DataType::BFLOAT16 || tensor.get_dtype() == DataType::FLOAT32,
+        "batch_norm only supports bfloat16, float32. Got: {}",
+        tensor.get_dtype());
+    TT_FATAL(
+        tensor.storage_type() == StorageType::DEVICE,
+        "Operands to batch_norm need to be on device! Got: {}",
+        tensor.storage_type());
+    TT_FATAL(tensor.buffer() != nullptr, "Operands to batch_norm need to be allocated in buffers on device!");
+}
```
Comment (patrickroberts) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/device/batch_norm_device_operation.cpp:25]:
```suggestion
inline void check_tensor_BN(const Tensor& tensor, std::string_view name, std::uint32_t input_c_dim) {
    TT_FATAL(
        tensor.get_layout() == Layout::TILE, "batch_norm only supports tiled layout. Got: {}", tensor.get_layout());
    TT_FATAL(
        tensor.get_dtype() == DataType::BFLOAT16 || tensor.get_dtype() == DataType::FLOAT32,
        "batch_norm only supports bfloat16, float32. Got: {}",
        tensor.get_dtype());
    TT_FATAL(
        tensor.storage_type() == StorageType::DEVICE,
        "Operands to batch_norm need to be on device! Got: {}",
        tensor.storage_type());
    TT_FATAL(tensor.buffer() != nullptr, "Operands to batch_norm need to be allocated in buffers on device!");
    TT_FATAL(tensor.get_logical_shape()[1] == input_c_dim, "{}[1] must be the same as input's channel size.", name);
}
```

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/batch_norm_device_operation.cpp [hunk 10]
```diff
@@ -8,21 +8,36 @@
...
+    check_tensor_BN(batch_mean);
+    check_tensor_BN(batch_var);
 
     // input (N, C, H, W)
     auto C = input.get_logical_shape()[1];
     // output (N, C, H, W)
     if (output.has_value()) {
+        check_tensor_BN(output.value());
         auto check_C = output.value().get_logical_shape()[1];
         TT_FATAL(C == check_C, "output_shape[1] must be the same as input's channel size.");
```
Comment (patrickroberts) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/device/batch_norm_device_operation.cpp:42]:
```suggestion
        check_tensor_BN(output.value(), "output_shape", C);
```

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/batch_norm_device_operation.cpp [hunk 11]
```diff
@@ -34,6 +49,7 @@ void BatchNormOperation::validate_tensors(
 
     // weight (1, C, 1, 1)
     if (weight.has_value()) {
+        check_tensor_BN(weight.value());
         TT_FATAL(
             weight.value().get_logical_shape()[1] == C, "weight_shape[1] must be the same as input's channel size.");
```
Comment (patrickroberts) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/device/batch_norm_device_operation.cpp:54]:
```suggestion
        check_tensor_BN(weight.value(), "weight_shape", C);
```
and please also remove the duplicate `TT_FATAL` below (the review tool won't let me include that line in my selection)

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/batch_norm_device_operation.cpp [hunk 12]
```diff
@@ -42,6 +58,7 @@ void BatchNormOperation::validate_tensors(
 
     // bias (1, C, 1, 1)
     if (bias.has_value()) {
+        check_tensor_BN(bias.value());
         TT_FATAL(bias.value().get_logical_shape()[1] == C, "bias_shape[1] must be the same as input's channel size.");
         TT_FATAL(bias.value().get_logical_shape()[1] == C, "bias_shape[1] must be the same as input's channel size.");
```
Comment (patrickroberts) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/device/batch_norm_device_operation.cpp:63]:
```suggestion
        check_tensor_BN(bias.value(), "bias_shape", C);
```

**Review by patrickroberts — CHANGES_REQUESTED** [review]

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/batch_norm_device_operation.cpp (hunk 9 above)
Comment (VirdhatchaniKN) [reply ttnn/cpp/ttnn/operations/normalization/batch_norm/device/batch_norm_device_operation.cpp:25]: Will add a condition for the same. Thanks for pointing out. 
Will resolve this shape constraint in the upcoming PR

**Review by patrickroberts — APPROVED** [review]

**Review by sjameelTT — APPROVED** [review]

**Review by dmakoviichuk-tt — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/compute/batch_norm_sfpu_kernel.cpp [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+    constexpr auto cb_batch_mean = tt::CBIndex::c_1;  // batch_mean
+    constexpr auto cb_output_0 =
+        tt::CBIndex::c_2;  // output -- > [(input - batch_mean)/(sqrt(batch_var + eps))] * weight
+    constexpr auto cb_batch_var = tt::CBIndex::c_3;  // batch_var
+    constexpr auto cb_eps = tt::CBIndex::c_4;        // eps
+    constexpr auto cb_den = tt::CBIndex::c_5;        // 1/(sqrt(batch_var + eps))
+    constexpr auto cb_num = tt::CBIndex::c_6;        // input - batch_mean
+    constexpr auto cb_weight = tt::CBIndex::c_16;    // weight tensor
+    constexpr auto cb_tmp_1 = tt::CBIndex::c_17;     // (input - batch_mean)/(sqrt(batch_var + eps))
+    constexpr auto cb_bias = tt::CBIndex::c_18;      // bias tensor
```
Comment (tt-aho) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/compute/batch_norm_sfpu_kernel.cpp:196]: This op needs to be updated to not hardcode cb ids in kernels, and have it passed in as compile time args. The ids should also be sequential from 0, not sparse, ie. cb 16, 17, 18 should be changed to 7, 8, 9. You can do this in a follow up pr if you want.

**Review by tt-aho — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/compute/batch_norm_sfpu_kernel.cpp (hunk 13 above)
Comment (VirdhatchaniKN) [reply ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/compute/batch_norm_sfpu_kernel.cpp:196]: Sure

**Review by mcw-anasuya — APPROVED** [review]

**Review by bbradelTT — APPROVED** [review]

**Review by dmakoviichuk-tt — APPROVED** [review]

**Review by dmakoviichuk-tt — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/compute/batch_norm_sfpu_kernel.cpp (hunk 13 above)
Comment (cmaryanTT) [reply ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/compute/batch_norm_sfpu_kernel.cpp:196]: @VirdhatchaniKN - please open a new issue to track this if you're going to do this in another PR

Comment (VirdhatchaniKN) [reply ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/compute/batch_norm_sfpu_kernel.cpp:196]: https://github.com/tenstorrent/tt-metal/issues/17758 - created a new issue to track this down

**Comment by VirdhatchaniKN** [comment]
https://github.com/tenstorrent/tt-metal/actions/runs/13217558701 - passed
